source/CCustomOpcodeSystem.cpp

```diff
@@ -992,6 +954,110 @@ namespace CLEO
+		if (returnArgs)
+		{
+			if (returnArgCount > 32)
+			{
+				SHOW_ERROR("Opcode [%04X] has too many (%d) args in script %s\nScript suspended.", opcode, returnArgCount, cs->GetInfoStr().c_str());
+				return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+			}
+
+			auto nVarArg = GetVarArgCount(thread);
+			if (returnArgCount > nVarArg)
+			{
+				SHOW_ERROR("Opcode [%04X] declared %d args, but %d was provided in script %s\nScript suspended.", opcode, returnArgCount, nVarArg, ((CCustomScript*)thread)->GetInfoStr().c_str());
+				return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+			}
+
+			for (DWORD i = 0; i < returnArgCount; i++)
+			{
+				SCRIPT_VAR* arg = arguments + i;
+
+				auto paramType = (eDataType)*thread->GetBytePointer();
+				if (IsImmInteger(paramType) || IsVariable(paramType))
+				{
+					*thread >> arg->dwParam;
+				}
+				else if (paramType == DT_FLOAT)
+				{
+					*thread >> arg->fParam;
+				}
+				else if (IsVarString(paramType))
+				{
+					auto str = ReadStringParam(thread); OPCODE_VALIDATE_STR_ARG_READ(str)
+					arg->pcParam = m_cleoReturnStr.Store(str);
+				}
+				else if (IsImmString(paramType)) // those texts exists in script code, but without terminator character. Copy is necessary
+				{
+					auto str = ReadStringParam(thread); OPCODE_VALIDATE_STR_ARG_READ(str)
+					arg->pcParam = m_cleoReturnStr.Store(str);
+				}
+				else
+				{
+					SHOW_ERROR("Invalid argument type '0x%02X' in opcode [%04X] in script %s\nScript suspended.", paramType, opcode, ((CCustomScript*)thread)->GetInfoStr().c_str());
+					return CCustomOpcodeSystem::ErrorSuspendScript(thread);
+				}
+			}
+		}
+
+		// handle program flow
+		scmFunc->Return(cs); // jump back to cleo_call, right after last input param. Return slot var args starts here
+		if (scmFunc->moduleExportRef != nullptr) GetInstance().ModuleSystem.ReleaseModuleRef((char*)scmFunc->moduleExportRef); // exiting export - release module
+		delete scmFunc;
+
+		if (returnArgs)
+		{
+			DWORD returnSlotCount = GetVarArgCount(cs);
+			if (returnSlotCount > returnArgCount || (strictArgCount && returnSlotCount < returnArgCount))
+			{
+				SHOW_ERROR("Opcode [%04X] returned %d params, while function caller expected %d in script %s\nScript suspended.", opcode, returnArgCount, returnSlotCount, cs->GetInfoStr().c_str());
+				return CCustomOpcodeSystem::ErrorSuspendScript(cs);
+			}
+			else if (returnSlotCount < returnArgCount)
+			{
+				LOG_WARNING(thread, "Opcode [%04X] returned %d params, while function caller expected %d in script %s", opcode, returnArgCount, returnSlotCount, cs->GetInfoStr().c_str());
+			}
+
+			// set return args
+			for (DWORD i = 0; i < returnArgCount; i++)
+			{
+				auto arg = (SCRIPT_VAR*)thread->GetBytePointer();
+
+				auto paramType = *(eDataType*)arg;
+				if (IsVariable(paramType))
+				{
+					*thread << arguments[i].dwParam;
+				}
+				else if (IsVarString(paramType))
+				{
+					WriteStringParam(thread, arguments[i].pcParam);

```

Comment:
the limit for a returned string is still 16 characters? if we return a longer string it would overwrite next variable?

Reply:
WriteStringParam respects target parameter type, is also handles null terminator.

Reply:
```
function foo
return "this is very long string that exceeds 16 char limit for one @v variable"
end

4@ = 0
0@v = foo
```

what the value of 0@v and will 4@ remain 0?

Reply:
0@ will contain 15 characters + terminator, 4@ is unchanged

Reply:
I don't think we should add null-terminator for "". If I need a null-terminated string, then I should use single quote. ''